File: notebooksjava/_09PalindromeLinkedList.java
import java.util.ArrayList;
import java.util.List;

public class _09PalindromeLinkedList {

    static class ListNode {
        int val;
        ListNode next;
        ListNode(int val) { this.val = val; }
    }

    // ノード値を配列に収集し、双ポインタで両端から比較する
    public boolean isPalindrome(ListNode head) {
        List<Integer> vals = new ArrayList<>();
        ListNode curr = head;
        while (curr != null) {
            vals.add(curr.val);
            curr = curr.next;
        }

        int left = 0, right = vals.size() - 1;
        while (left < right) {
            if (!vals.get(left).equals(vals.get(right))) return false;
            left++;
            right--;
        }
        return true;
    }

    // ── 補助関数 ──────────────────────────────────────────────

    static ListNode createList(int... vals) {
        if (vals.length == 0) return null;
        ListNode head = new ListNode(vals[0]);
        ListNode curr = head;
        for (int i = 1; i < vals.length; i++) {
            curr.next = new ListNode(vals[i]);
            curr = curr.next;
        }
        return head;
    }

    public static void main(String[] args) {
        _09PalindromeLinkedList sol = new _09PalindromeLinkedList();

        assert sol.isPalindrome(null)                      == true  : "null → true";
        assert sol.isPalindrome(createList(7))             == true  : "[7] → true";
        assert sol.isPalindrome(createList(1, 2, 1))       == true  : "[1,2,1] → true";
        assert sol.isPalindrome(createList(1, 2, 2, 1))    == true  : "[1,2,2,1] → true";
        assert sol.isPalindrome(createList(1, 2, 3, 1))    == false : "[1,2,3,1] → false";
        assert sol.isPalindrome(createList(1, 2))          == false : "[1,2] → false";

        System.out.println("All test cases passed ✅");
    }
}

File: notebooksc#/09-palindrome_linked_list.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;

public class ListNode
{
    public int Val { get; set; }
    public ListNode Next { get; set; }

    public ListNode(int val = 0, ListNode next = null)
    {
        Val = val;
        Next = next;
    }
}

public class Solution09PalindromeLinkedList
{
    // Check if linked list is palindrome by collecting values in a list
    // and checking from both ends
    public bool IsPalindrome(ListNode head)
    {
        List<int> vals = new List<int>();
        ListNode curr = head;

        while (curr != null)
        {
            vals.Add(curr.Val);
            curr = curr.Next;
        }

        int left = 0;
        int right = vals.Count - 1;

        while (left < right)
        {
            if (vals[left] != vals[right])
                return false;
            left++;
            right--;
        }

        return true;
    }

    // Helper function to create a linked list from variable arguments
    public ListNode CreateList(params int[] vals)
    {
        if (vals.Length == 0)
            return null;

        ListNode head = new ListNode(vals[0]);
        ListNode curr = head;

        for (int i = 1; i < vals.Length; i++)
        {
            curr.Next = new ListNode(vals[i]);
            curr = curr.Next;
        }

        return head;
    }

    static void Main()
    {
        var sol = new Solution09PalindromeLinkedList();

        // Test case 1: null list
        Debug.Assert(sol.IsPalindrome(null) == true, "null → true");

        // Test case 2: single element
        Debug.Assert(sol.IsPalindrome(sol.CreateList(7)) == true, "[7] → true");

        // Test case 3: odd length palindrome
        Debug.Assert(sol.IsPalindrome(sol.CreateList(1, 2, 1)) == true, "[1,2,1] → true");

        // Test case 4: even length palindrome
        Debug.Assert(sol.IsPalindrome(sol.CreateList(1, 2, 2, 1)) == true, "[1,2,2,1] → true");

        // Test case 5: not a palindrome (odd length)
        Debug.Assert(sol.IsPalindrome(sol.CreateList(1, 2, 3, 1)) == false, "[1,2,3,1] → false");

        // Test case 6: not a palindrome (even length)
        Debug.Assert(sol.IsPalindrome(sol.CreateList(1, 2)) == false, "[1,2] → false");

        // Test case 7: longer palindrome
        Debug.Assert(sol.IsPalindrome(sol.CreateList(1, 2, 3, 2, 1)) == true, "[1,2,3,2,1] → true");

        Console.WriteLine("All test cases passed ✅");
    }
}
